src/S3IteratorLibsvm.cpp

```diff
@@ -0,0 +1,412 @@
+  if (random_access) {
+    srand(42 + worker_id);
+  }
+}
+
+std::shared_ptr<SparseDataset> S3IteratorLibsvm::getNext() {
+  sem_wait(&semaphore);
+  ring_lock.lock();
+
+  // first discard empty queue
+  while (minibatches_list.front().size() == 0) {
+    auto queue_ptr = minibatches_list.pop();
+  }
+  auto ret = minibatches_list.front().back();
+  minibatches_list.front().pop_back();
+  num_minibatches_ready--;
+  ring_lock.unlock();
+
+  // FIXME this should be calculating the local amount of memory
+  if (num_minibatches_ready < 200 && pref_sem.getvalue() < (int) read_ahead) {
+#ifdef DEBUG
+    std::cout << "getNext::pref_sem.signal" << std::endl;
+#endif
+    pref_sem.signal();
+  }
+
+  return ret;
+}
+
+/**
+ * Moves index forward while data[index] is a space
+ * returns true if it ended on a digit, otherwise returns false
+ */
+bool S3IteratorLibsvm::ignoreSpaces(uint64_t& index, const std::string& data) {
+  while (isspace(data[index])) {
+    index++;
+  }
+  return isdigit(data[index]);
+}
+
+bool S3IteratorLibsvm::ignoreSpacesNotNewline(uint64_t& index,
+                                              const std::string& data) {
+  while (data[index] != '\n' && isspace(data[index])) {
+    index++;
+  }
+  return isdigit(data[index]);
+}
+
+template <class T>
+T S3IteratorLibsvm::readNum(uint64_t& index, std::string& data) {
+  if (!isdigit(data[index])) {
+    throw std::runtime_error("Error in the dataset");
+  }
+
+  uint64_t index_fw = index;
+  // if float or double also accept dots
+  if constexpr (std::is_same<T, double>::value ||
+                std::is_same<T, float>::value) {
+    while (isdigit(data[index_fw]) || data[index_fw] == '.') {
+      index_fw++;
+    }
+  } else {
+    while (isdigit(data[index_fw])) {
+      index_fw++;
+    }
+  }
+
+  char c = data[index_fw];
+  data[index_fw] = 0;
+
+  T result;
+  if constexpr (std::is_same<T, int>::value) {
+    sscanf(&data[index], "%d", &result);
+  } else if constexpr (std::is_same<T, double>::value) {
+    sscanf(&data[index], "%lf", &result);
+  } else if constexpr (std::is_same<T, float>::value) {
+    sscanf(&data[index], "%f", &result);
+  } else if constexpr (std::is_same<T, uint64_t>::value) {
+    sscanf(&data[index], "%lu", &result);
+  } else {
+    throw std::runtime_error("Data type not supported");
+  }
+
+  data[index_fw] = c;  // repair
+  index = index_fw;
+  return result;
+}
+
+bool S3IteratorLibsvm::buildDatasetCsv(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+bool S3IteratorLibsvm::buildDatasetVowpalWabbit(
+    const std::string& data,
+    uint64_t index,
+    std::shared_ptr<SparseDataset>& minibatch) {
+  return false;
+}
+
+/**
+ * Build minibatch from text in libsvm format
+ * We assume index is at a start of a line
+ * index moves forward
+ */
+bool S3IteratorLibsvm::buildDatasetLibsvm(
+    std::string& data,
```

Comment:
Declare all data parameters as const, since we don't want to accidentally change these.